File: src/main/java/lv/ldn/scalaoop/java/DrillDown.java
package lv.ldn.scalaoop.java;

import java.text.SimpleDateFormat;
import java.util.Date;

public final class DrillDown {

    private DrillDown() {
    }

    interface Server {
        Disk getDisk();
    }

    interface Disk {
        Snapshot getSnapshot();

        int getCapacity();
    }

    interface Snapshot {
        Date getDate();
    }

    public String getSnapshotDate(Server server) {
        Disk disk = server.getDisk();
        if (disk != null) {
            Snapshot snapshot = disk.getSnapshot();
            if (snapshot != null) {
                Date date = snapshot.getDate();
                if (date != null) {
                    return new SimpleDateFormat("YYYY-mm-dd").format(date);
                }
            }
        }
        return "(none)";
    }

}

File: src/main/scala/lv/ldn/scalaoop/scala/DrillDown.scala
package lv.ldn.scalaoop.scala

import java.util.Date
import java.text.SimpleDateFormat


trait Server {
  val disk: Option[Disk]
}

trait Disk {
  val snapshot: Option[Snapshot]

  val capacity: Int
}

trait Snapshot {
  val date: Option[Date]
}

object DrillDown {
  def getSnapshotDate(server: Server): String = {
    val maybeDate = for {
      disk <- server.disk
      snapshot <- disk.snapshot
      date <- snapshot.date
    } yield new SimpleDateFormat("YYYY-mm-dd").format(date)
    maybeDate.getOrElse("(none)")
  }
}
